File: Windows_csharp/Node.cs
/*
 * Title:  spider Node.cs (Windows csharp)
 * Author: Shuichiro Endo
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Net.Sockets;

namespace spider
{
    public class Node
    {
        protected const int NODE_BUFFER_SIZE = 72000;
        protected const int SPIDER_MESSAGE_DATA_SIZE = 65536;
        protected const int SPIDER_MESSAGE_DATA_MAX_SIZE = 65552;     // 65536 + 16 (AES padding)
        protected const int INET6_ADDR_STRING_LENGTH = 46;

        protected TcpListener tcpListener = null;
        protected TcpClient tcpClient = null;
        protected UdpClient udpClient = null;
        protected IntPtr sock = IntPtr.Zero;
        protected NetworkStream stream = null;
        protected MessageManager messageManager;
        protected Socks5MessageQueue socks5MessagesQueue;
        protected Dictionary<uint, Socks5Message> socks5ReceiveMessagesMap = new Dictionary<uint, Socks5Message>();
        protected static readonly object socks5ReceiveMessagesMapLock = new object();

        public Node(TcpListener tcpListener,
                    TcpClient tcpClient,
                    MessageManager messageManager)
        {
            this.tcpListener = tcpListener;
            if(this.tcpListener != null)
            {
                this.sock = this.tcpListener.Server.Handle;
            }
            this.tcpClient = tcpClient;
            if(this.tcpClient != null)
            {
                this.sock = this.tcpClient.Client.Handle;
                this.stream = this.tcpClient.GetStream();
            }
            this.MessageManager = messageManager;
            this.socks5MessagesQueue = new Socks5MessageQueue();
        }

        public Node(UdpClient udpClient,
                    MessageManager messageManager)
        {
            this.udpClient = udpClient;
            if(this.udpClient != null)
            {
                this.sock = this.udpClient.Client.Handle;
            }
            this.MessageManager = messageManager;
            this.socks5MessagesQueue = new Socks5MessageQueue();
        }

        ~Node()
        {

        }

        public TcpListener TcpListener
        {
            get { return tcpListener; }
            set { tcpListener = value; }
        }

        public TcpClient TcpClient
        {
            get { return tcpClient; }
            set { tcpClient = value; }
        }

        public UdpClient UdpClient
        {
            get { return udpClient; }
            set { udpClient = value; }
        }

        public IntPtr Sock
        {
            get { return sock; }
            set { sock = value; }
        }

        public NetworkStream Stream
        {
            get { return stream; }
            set { stream = value; }
        }

        public MessageManager MessageManager
        {
            get { return messageManager; }
            set { messageManager = value; }
        }

        public void PushSocks5Message(Socks5Message socks5Message)
        {
            socks5MessagesQueue.Push(socks5Message);

            return;
        }

        public void PushSocks5Message(Object obj)
        {
            Socks5Message socks5Message = obj as Socks5Message;

            if(socks5Message != null)
            {
                socks5MessagesQueue.Push(socks5Message);
            }

            return;
        }

        public int PushTimeoutSocks5Message(Socks5Message socks5Message,
                                            int tvSec,
                                            int tvUsec)
        {
            int ret = 0;

            ret = socks5MessagesQueue.PushTimeout(socks5Message,
                                                  tvSec,
                                                  tvUsec);

            return ret;
        }

        protected Socks5Message PopSocks5Message()
        {
            Socks5Message socks5Message = socks5MessagesQueue.Pop();

            return socks5Message;
        }

        protected Socks5Message PopTimeoutSocks5Message(int tvSec,
                                                        int tvUsec)
        {
            Socks5Message socks5Message = socks5MessagesQueue.PopTimeout(tvSec,
                                                                         tvUsec);

            return socks5Message;
        }

        public void AddSocks5ReceiveMessageToMap(Object obj)
        {
            Socks5Message socks5Message = obj as Socks5Message;
            uint message_id = 0;

            if(socks5Message != null)
            {
                message_id = socks5Message.MessageId;

                lock(socks5ReceiveMessagesMapLock)
                {
                    socks5ReceiveMessagesMap.Add(message_id,
                                                 socks5Message);
                }
            }

            return;
        }

        protected Socks5Message GetSocks5ReceiveMessageFromMap(uint messageId)
        {
            Socks5Message socks5Message = null;

            lock(socks5ReceiveMessagesMapLock)
            {
                if(socks5ReceiveMessagesMap.ContainsKey(messageId))
                {
                    socks5Message = socks5ReceiveMessagesMap[messageId];
                    socks5ReceiveMessagesMap.Remove(messageId);
                }
            }

            return socks5Message;
        }

        public int RecvData(byte[] buffer,
                            int bufferSize,
                            int tvSec,
                            int tvUsec)
        {
            int rec = 0;
            tcpClient.ReceiveTimeout = tvSec * 1000 + tvUsec / 1000;

            if(bufferSize > NODE_BUFFER_SIZE)
            {
#if DEBUGPRINT
                Console.WriteLine("[-] buffer size error");
#endif
                return -1;
            }

            try
            {
                rec = stream.Read(buffer,
                                  0,
                                  NODE_BUFFER_SIZE);
            }catch(IOException ex)
            {
#if DEBUGPRINT
                Console.WriteLine("[-] RecvData timeout: {0}",
                                  ex.Message);
#endif
                return -1;
            }catch (Exception ex)
            {
#if DEBUGPRINT
                Console.WriteLine("[-] RecvData error: {0}",
                                  ex.Message);
#endif
                return -1;
            }

            return rec;
        }

        public int SendData(byte[] buffer,
                            int dataSize,
                            int tvSec,
                            int tvUsec)
        {
            int sen = 0;
            this.tcpClient.SendTimeout = tvSec * 1000 + tvUsec / 1000;

            if(dataSize > NODE_BUFFER_SIZE)
            {
#if DEBUGPRINT
                Console.WriteLine("[-] data size error");
#endif
                return -1;
            }

            try
            {
                stream.Write(buffer,
                             0,
                             dataSize);
            }catch(IOException ex)
            {
#if DEBUGPRINT
                Console.WriteLine("[-] SendData timeout: {0}",
                                  ex.Message);
#endif
                return -1;
            }catch (Exception ex)
            {
#if DEBUGPRINT
                Console.WriteLine("[-] SendData error: {0}",
                                  ex.Message);
#endif
                return -1;
            }

            sen = dataSize;

            return sen;
        }

        public uint GenerateRandomId()
        {
            Random random = new Random();
            ulong randomValue = 0;

            do
            {
                randomValue = (ulong)random.Next(1, int.MaxValue) + (ulong)random.Next(0, int.MaxValue);
            }while(randomValue >= (ulong)uint.MaxValue);

            return (uint)randomValue;
        }

        public ushort NetworkToHostOrderUShort(ushort value)
        {
            if(BitConverter.IsLittleEndian)
            {
                byte[] bytes = BitConverter.GetBytes(value);
                Array.Reverse(bytes);
                return BitConverter.ToUInt16(bytes, 0);
            }

            return value;
        }

        public ushort HostToNetworkOrderUShort(ushort value)
        {
            if(BitConverter.IsLittleEndian)
            {
                byte[] bytes = BitConverter.GetBytes(value);
                Array.Reverse(bytes);
                return BitConverter.ToUInt16(bytes, 0);
            }

            return value;
        }

        public void PrintBytes(byte[] buffer,
                               int bufferLength)
        {
            for(int i = 0; i < bufferLength; i++)
            {
                if(i != 0 && i % 16 == 0){
                    Console.WriteLine("");
                }else if(i % 16 == 8)
                {
                    Console.Write(" ");
                }
                Console.Write($"{buffer[i]:X2} ");
            }
            Console.WriteLine("");
        }
    }
}

File: Linux/node.cpp
/*
 * Title:  spider node.cpp (Linux)
 * Author: Shuichiro Endo
 */

#include "spider.hpp"
#include "node.hpp"
#include "messagemanager.hpp"
#include "socks5message.hpp"
#include "socks5messagequeue.hpp"


namespace spider
{
    Node::Node(int32_t sock,
              std::shared_ptr<Messagemanager> message_manager)
    {
        this->sock = sock;
        this->message_manager = message_manager;
        this->socks5_messages_queue = std::make_unique<Socks5messagequeue>();
    }

    Node::~Node()
    {

    }

    void Node::set_sock(int32_t sock)
    {
        this->sock = sock;
    }

    int32_t Node::get_sock()
    {
        return sock;
    }

    void Node::set_message_manager(std::shared_ptr<Messagemanager> messagemanager)
    {
        this->message_manager = message_manager;
    }

    std::shared_ptr<Messagemanager> Node::get_message_manager()
    {
        return message_manager;
    }

    void Node::push_socks5_message(std::shared_ptr<Socks5message> socks5_message)
    {

        socks5_messages_queue->push(socks5_message);

        return;
    }

    int32_t Node::push_timeout_socks5_message(std::shared_ptr<Socks5message> socks5_message,
                                              int32_t tv_sec,
                                              int32_t tv_usec)
    {
        int ret = 0;

        ret = socks5_messages_queue->push_timeout(socks5_message,
                                                  tv_sec,
                                                  tv_usec);

        return ret;
    }

    std::shared_ptr<Socks5message> Node::pop_socks5_message()
    {
        std::shared_ptr<Socks5message> socks5_message = socks5_messages_queue->pop();

        return socks5_message;
    }

    std::shared_ptr<Socks5message> Node::pop_timeout_socks5_message(int32_t tv_sec,
                                                                    int32_t tv_usec)
    {
        std::shared_ptr<Socks5message> socks5_message = socks5_messages_queue->pop_timeout(tv_sec,
                                                                                           tv_usec);

        return socks5_message;
    }

    void Node::add_socks5_receive_message_to_map(uint32_t message_id,
                                                 std::shared_ptr<Socks5message> socks5_message)
    {
        std::unique_lock<std::mutex> lock(socks5_receive_messages_map_mutex);

        socks5_receive_messages_map.insert({message_id, socks5_message});

        lock.unlock();

        return;
    }

    std::shared_ptr<Socks5message> Node::get_socks5_receive_message_from_map(uint32_t message_id)
    {
        std::shared_ptr<Socks5message> socks5_message = nullptr;
        std::size_t check_key_count = 0;

        std::unique_lock<std::mutex> lock(socks5_receive_messages_map_mutex);

        check_key_count = socks5_receive_messages_map.count(message_id);
        if(check_key_count > 0)
        {
            socks5_message = socks5_receive_messages_map[message_id];
            socks5_receive_messages_map.erase(message_id);
        }

        lock.unlock();

        return socks5_message;
    }

    int32_t Node::recv_data(char *buffer,
                            int32_t buffer_size,
                            long tv_sec,
                            long tv_usec)
    {
        int ret = 0;
        int32_t rec = 0;
        fd_set readfds;
        int nfds = -1;
        struct timeval tv;


        if(buffer_size > NODE_BUFFER_SIZE)
        {
#ifdef _DEBUG
            std::printf("[-] buffer size error\n");
#endif
            return -1;
        }

        std::memset((char *)buffer,
                    0,
                    buffer_size);

        while(1)
        {
            FD_ZERO(&readfds);
            FD_SET(sock,
                   &readfds);
            nfds = sock + 1;
            tv.tv_sec = tv_sec;
            tv.tv_usec = tv_usec;

            ret = select(nfds,
                         &readfds,
                         NULL,
                         NULL,
                         &tv);
            if(ret == 0)
            {
#ifdef _DEBUG
                std::printf("[-] recv_data select timeout\n");
#endif
                return -1;
            }

            ret = FD_ISSET(sock,
                           &readfds);
            if(ret)
            {
                rec = recv(sock,
                           buffer,
                           NODE_BUFFER_SIZE,
                           0);
                if(rec <= 0)
                {
                    if(errno == EINTR)
                    {
                        continue;
                    }else if(errno == EAGAIN)
                    {
                        std::this_thread::sleep_for(std::chrono::microseconds(5000));
                        continue;
                    }else
                    {
#ifdef _DEBUG
                        std::printf("[-] recv error: %d\n",
                                    errno);
#endif
                        return -1;
                    }
                }else
                {
                    break;
                }
            }
        }

        return rec;
    }

    int32_t Node::send_data(char *buffer,
                            int32_t data_size,
                            long tv_sec,
                            long tv_usec)
    {
        int ret = 0;
        int32_t sen = 0;
        int32_t send_length = 0;
        int32_t len = 0;
        fd_set writefds;
        int nfds = -1;
        struct timeval tv;


        if(data_size > NODE_BUFFER_SIZE)
        {
#ifdef _DEBUG
            std::printf("[-] data size error\n");
#endif
            return -1;
        }

        len = data_size;

        while(len > 0)
        {
            FD_ZERO(&writefds);
            FD_SET(sock,
                   &writefds);
            nfds = sock + 1;
            tv.tv_sec = tv_sec;
            tv.tv_usec = tv_usec;

            ret = select(nfds,
                         NULL,
                         &writefds,
                         NULL,
                         &tv);
            if(ret == 0){
#ifdef _DEBUG
                std::printf("[-] send_data select timeout\n");
#endif
                return -1;
            }

            ret = FD_ISSET(sock,
                           &writefds);
            if(ret)
            {
                sen = send(sock,
                           buffer+send_length,
                           len,
                           MSG_NOSIGNAL);
                if(sen <= 0)
                {
                    if(errno == EINTR)
                    {
                        continue;
                    }else if(errno == EAGAIN)
                    {
                        std::this_thread::sleep_for(std::chrono::microseconds(5000));
                        continue;
                    }else
                    {
#ifdef _DEBUG
                        std::printf("[-] send error: %d\n",
                                    errno);
#endif
                        return -1;
                    }
                }
                send_length += sen;
                len -= sen;
            }
        }

        return data_size;
    }

    uint32_t Node::generate_random_id()
    {
        std::random_device rd;
        std::mt19937 mt(rd());
        std::uniform_int_distribution<uint32_t> distrib(1, UINT32_MAX);

        return distrib(mt);
    }

    void Node::print_bytes(char *buffer, int buffer_length)
    {
        for(int i=0; i<buffer_length; i++){
            if(i != 0 && i%16 == 0){
                std::printf("\n");
            }else if(i%16 == 8){
                std::printf(" ");
            }
            std::printf("%02x ", buffer[i] & 0xff);
        }
        std::printf("\n");

        return;
    }
}

